ts/types.go

```diff
@@ -21,14 +21,69 @@
 package ts
 
 import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
 	"time"
+	"unsafe"
 )
 
+// ID represents an immutable identifier for a timeseries
+type ID interface {
+	fmt.Stringer
+
+	Data() []byte
+	Hash() Hash
+}
+
+// Hash represents a form of ID suitable to be used as map keys
+type Hash [md5.Size]byte
+
+var null = Hash{}
+
+// Fold folds the hash ID to a 64-bit unsigned integer.
+func (h Hash) Fold() uint64 {
+	return *(*uint64)(unsafe.Pointer(&h[0])) ^ *(*uint64)(unsafe.Pointer(&h[8]))
+}
+
+type id struct {
+	data []byte
+	hash Hash
+}
+
+// Data returns the binary representation of an ID
+func (id *id) Data() []byte {
+	return id.data
+}
+
+// Hash returns (and calculates, if needed) the hash for an ID
+func (id *id) Hash() Hash {
+	if bytes.Equal(id.hash[:], null[:]) {
+		id.hash = md5.Sum(id.data)
```

Comment:
I'm not so sure about it. I can't really find anything about it being specifically resistant to collisions. I'd rather go with a conservative MD5 here for now and see how it behaves in production.